Add API for CreakingHeart

Doc94 opened the pull request:
Add more methods to the Creaking Heart

Doc94 committed:
add methods for CreakingHeart

- allow manage creaking attached
- helper method for spawn creaking

Doc94 committed:
missing behaviour for make the creaking protect the block

Doc94 committed:
fix typo in precondition

Doc94 committed:
add CreakingHeart#spreadResin()

Doc94 committed:
add methods for max distance of Creaking attached

Doc94 committed:
handle spawn with the limits of distance for avoid bucle of spawn-death for low values

notTamion commented on paper-api/src/main/java/org/bukkit/block/CreakingHeart.java:
```diff
@@ -1,7 +1,61 @@
+     *
+     * @return the {@link Creaking} for protect the creaking heart or null if fails
+     */
+    @NullMarked
```
Wrong annotation

Doc94 replied:
IJ try to make me use this like 4 times.. and forget that last xd fixed...

notTamion commented on paper-api/src/main/java/org/bukkit/block/CreakingHeart.java:
```diff
@@ -1,7 +1,61 @@
+     */
+    @Nullable
+    Location spreadResin();
+
```
Extra new line

notTamion commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,56 @@
+         BlockPos blockPos = creakingHeart.getBlockPos();
+         Optional<Creaking> optional = SpawnUtil.trySpawnMob(
+-            EntityType.CREAKING, EntitySpawnReason.SPAWNER, level, blockPos, 5, 16, 8, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER_NO_LEAVES, true
++            EntityType.CREAKING, EntitySpawnReason.SPAWNER, level, blockPos, 5, Math.min(SPAWN_RANGE_XZ, creakingHeart.getDistanceCreakingTooFar()), Math.min(SPAWN_RANGE_Y, creakingHeart.getDistanceCreakingTooFar()), SpawnUtil.Strategy.ON_TOP_OF_COLLIDER_NO_LEAVES, true // Paper - handle too far distance for avoid spawn out of limits
```
Not sure why we need to min this? i think we should also increase the spawn range when we increase the max Distance.

EDIT: or add extra api for that

Doc94 replied:
yeah not sure how handle this... the idea its just avoid if i set the too far distance to 2, the block not turns into a "spawn creaking, its far, break, try again, its far, try again" any suggest?

notTamion replied:
Well that isn‘t going to be a problem is we scale the spawn range according to the maxDistance. So i think the best thing to do is either add extra api for the spawn range (though i think thats a bit excessive) or change the spawn range by the same factor as maxDistance was changed. but the current behaivour for this is sort of weird

Doc94 replied:
yeah still a little lost in this... unless just restrict the distance for not being low than the range... but dont like that... then not sure how best can modify here.

notTamion replied:
I was thinking about something along the lines of this:

```suggestion
+            EntityType.CREAKING, EntitySpawnReason.SPAWNER, level, blockPos, 5, creakingHeart.getDistanceCreakingTooFar() / creakingHeart.DISTANCE_CREAKING_TOO_FAR * SPAWN_RANGE_XZ, creakingHeart.getDistanceCreakingTooFar() / creakingHeart.DISTANCE_CREAKING_TOO_FARcreakingHeart.DISTANCE_CREAKING_TOO_FAR *SPAWN_RANGE_Y, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER_NO_LEAVES, true // Paper - handle too far distance for avoid spawn out of limits
```

lynxplay replied:
I'd be in favour of *potentially* adding extra API to increase the spawn radius down the line, for now I am unsure if we need this? The most important usecase I see for this is reducing the range.

I am fine with the min here to avoid the look of spawning one outside the rage and have it die immediately. 
Scaling it is a nice idea, but I think I'd prefer extra API.

Doc94 replied:
oh.. then what is better?
- the last change for the scaling (broken need fix that)
- the min thing and in another PR implement a proper api for that

lynxplay replied:
Ah, I personally am not much of a fan of *scaling* it, I think I'd just add the API for it.
Preconditions to verify that spawn <= survival range and survival range >= spawn range etc.
Scaling seems nice for auto vanilla-like feeling to the API but I think API just offers more specific control and doesn't "magically" change values.

Doc94 replied:
ok that logic currently just return 0 if try to set the distance in 2...

notTamion reviewed (changes requested):
some grammar in the jd needs to be fixed

Doc94 committed:
remove new line in final

Doc94 committed:
fix CreakingHeart#spawnCreaking() annotation

lynxplay commented on paper-api/src/main/java/org/bukkit/block/CreakingHeart.java:
```diff
@@ -1,7 +1,60 @@
+    /**
+     * Attempts to spawn a creaking for protect this creaking heart.
+     *
+     * @return the {@link Creaking} for protect the creaking heart or null if fails
```
I am a bit confused regarding the usage of this method.

Simulating the spawning behaviour of an entity type is something that might be useful API, but I don't think I'd want it to live in random block state types like this.
Developers can very well just spawn a creeking and set it via the setter.
The spawnProtector method really doesn't do more than spawning it + game events.

Doc94 replied:
hmmm yeah.. i fell useful just call the same NMS logic but maybe can just remove that and let users use the normal ways.

lynxplay commented on paper-api/src/main/java/org/bukkit/block/CreakingHeart.java:
```diff
@@ -1,7 +1,60 @@
+    /**
+     * Attempts to spread resin to adjacent blocks.
+     *
+     * @return the location of spread resin or null if it cannot spread
```
Should note that this might call events.

Doc94 committed:
update logic for spawn creaking protector in creaking heart

Co-authored-by: Tamion <70228790+notTamion@users.noreply.github.com>

Lulu13022002 commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,56 @@
+         } else {
+             this.clearCreakingInfo();
+         }
++        this.distanceCreakingTooFar = tag.contains(PAPER_CREAKING_MAX_DISTANCE_TAG, 99) ? tag.getInt(PAPER_CREAKING_MAX_DISTANCE_TAG) : DISTANCE_CREAKING_TOO_FAR; // Paper - Custom max distance for Creaking
```
Use the constants in Tag

lynxplay commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,56 @@
+         } else {
+             this.clearCreakingInfo();
+         }
++        this.distanceCreakingTooFar = tag.contains(PAPER_CREAKING_MAX_DISTANCE_TAG, 99) ? tag.getInt(PAPER_CREAKING_MAX_DISTANCE_TAG) : DISTANCE_CREAKING_TOO_FAR; // Paper - Custom max distance for Creaking
```
Use `Tag.TAG_ANY_NUMERIC` instead of 99.

Doc94 replied:
okay i base in the beacon class (then that class need use tha Tag class too) but fixed here.

Lulu13022002 reviewed (commented):
You should also handle unplaced block state properly

lynxplay commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,56 @@
+             tag.putUUID("creaking", this.creakingInfo.map(Entity::getUUID, uuid -> (UUID)uuid));
+         }
+-    }
++        tag.putInt(PAPER_CREAKING_MAX_DISTANCE_TAG, this.distanceCreakingTooFar); // Paper - Custom max distance for Creaking
```
A bit rough here if we wanna write.
We should keep the int but we'd ideal also want a way for the developer to fall back to vanillas value *even if it changes across versions*.

The current implementation would lock every creaking heart at the current vanilla value.

Doc94 replied:
Maybe just a 
`if (this.distanceCreakingTooFar != DISTANCE_CREAKING_TOO_FAR) tag.putInt(PAPER_CREAKING_MAX_DISTANCE_TAG, this.distanceCreakingTooFar); // Paper - Custom max distance for Creaking` ?

lynxplay commented on paper-api/src/main/java/org/bukkit/block/CreakingHeart.java:
```diff
@@ -1,7 +1,60 @@
+     * @return the creaking or null if this creaking heart don't have protector.
+     */
+    @Nullable
+    Creaking getCreaking();
```
Anything against sticking to the vanilla naming here? E.g. `getProtector`?
Other peoples input here before actually changing this tho.

Owen1212055 replied:
Get creaking makes more sense than protector, because people wouldn't assume that the creaking is "protecting" the heart but instead is spawned by it.

lynxplay commented on paper-api/src/main/java/org/bukkit/block/CreakingHeart.java:
```diff
@@ -1,7 +1,60 @@
+     *
+     * @return the max distance
+     */
+    int getMaxDistanceForCreaking();
```
I do not like this method name.
Especially when we also might add API for the creaking spawn distance.

`getProtectorSurvivalRange`
`setProtectorSurvivalRange`
`getProtectorSpawnRange`
`setProtectorSpawnRange`
*might* work, more peoples input tho pls.

Doc94 replied:
the lasts ones not like just becauuse that value its not for spawn.. its for limit the max distance where a creaking can "live"

maybe the first can work.. but not sure..

lynxplay replied:
The last ones were for the new API regarding the replacement for scaling, sorry xD
But yea, more people time chime in.

Owen1212055 replied:
getCreakingRemovalDistance?

Doc94 replied:
> getCreakingRemovalDistance?

not sounds bad... maybe can use that then (based in not another comments xd)

Doc94 committed:
Revert "update logic for spawn creaking protector in creaking heart"

This reverts commit 0306df1c739b94bebf64336b0aea1ee0f6917659.

Doc94 committed:
use correct tag definition and only save if value its not default vanilla

Doc94 committed:
add apinote to CreakingHeart#spreadResin() for events being called

Doc94 committed:
[ci skip] rename api method for removal distance

Doc94 committed:
Rename methods for Creaking remove distance

Doc94 committed:
Fix 1.21.5 renames

Doc94 committed:
Better check for null level in spawnCreaking

Lulu13022002 commented on paper-server/src/main/java/org/bukkit/craftbukkit/block/CraftCreakingHeart.java:
```diff
@@ -24,4 +33,51 @@ public CraftCreakingHeart copy() {
+
+    @Override
+    public void setCreakingRemovalDistance(final int distance) {
+        Preconditions.checkArgument(distance >= 0, "the distance must be >= 0");
```
You are allowing 0, yet the codec it serialize into only allow positive integer.

Lulu13022002 commented on paper-server/src/main/java/org/bukkit/craftbukkit/block/CraftCreakingHeart.java:
```diff
@@ -24,4 +33,51 @@ public CraftCreakingHeart copy() {
+    @Nullable
+    @Override
+    public Creaking spawnCreaking() {
+        if (!this.getBlockEntity().hasLevel()) {
```
Like i said previously, you should handle unplaced blockstate (calling requiredPlaced) for methods that need it. (This is not the only method that needs it)

Lulu13022002 commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,56 @@
+ 
+     @Nullable
+     public static Creaking spawnProtector(ServerLevel level, CreakingHeartBlockEntity creakingHeart) {
++        if (creakingHeart.getDistanceCreakingTooFar() <= 0) return null; // Paper - avoid spawn if the max distance is 0
```
Is this even possible now with the codec?

Doc94 replied:
if allow the 0 then yes.

Lulu13022002 commented on paper-server/patches/sources/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java.patch:
```diff
@@ -0,0 +1,56 @@
+     protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
+         super.loadAdditional(tag, registries);
+         tag.read("creaking", UUIDUtil.CODEC).ifPresentOrElse(this::setCreakingInfo, this::clearCreakingInfo);
++        tag.read(PAPER_CREAKING_MAX_DISTANCE_TAG, net.minecraft.util.ExtraCodecs.POSITIVE_INT).ifPresentOrElse(this::setDistanceCreakingTooFar, this::resetDistanceCreakingTooFar); // Paper - Custom max distance for Creaking
```
I don't think the methods are needed if you only had a public field then you could do smth like:
`distanceCreakingTooFar = tag.read(PAPER_CREAKING_MAX_DISTANCE_TAG, ExtraCodecs.POSITIVE_INT).orElse(DISTANCE_CREAKING_TOO_FAR)`

Doc94 replied:
okay i remove the methods and handle in the read with only the ifPresent and keep the field with the default value.

Lulu13022002 replied:
I think you still need to fallback to the default (when the tag is not there) to fully support the data command and reloads.

Doc94 committed:
Allow 0 for the CreakingRemovalDistance

Doc94 committed:
Use requirePlaced

Doc94 committed:
Reduce added methods in NMS patches

Doc94 committed:
Add fallback distanceCreakingTooFar when tag is missing

Doc94 committed:
CreakingHeartBlockEntity.java.patch fix for renames

Doc94 committed:
Merge branch 'main' of github.com:PaperMC/Paper into feature/creaking-heart-methods

Doc94 committed:
Fix spreadResin method signature

Doc94 committed:
Remove the reason why this was stuck in first place

Doc94 committed:
Handle changes since last update

Doc94 committed:
Merge branch 'main' of github.com:PaperMC/Paper into feature/creaking-heart-methods

Doc94 committed:
Cleanup for CraftCreakingHeart

- Missing Nullable annotation for construct a instance with not location (like happen in clone)
- Change how check the ServerLevel for avoid IDE warnings correctly

Warriorrrr committed:
minor tweaks

Owen1212055 reviewed (approved):
The only thing I am unsure of is manually playing the sounds, I would think its fine without. Other than that, lgtm

Doc94 committed:
Remove sound...

Warriorrrr committed:
Grammar pass

Warriorrrr committed:
Merge branch 'main' into feature/creaking-heart-methods

Warriorrrr reviewed (approved):
Re-approving to trigger workflow again

Warriorrrr committed:
Merge branch 'main' into feature/creaking-heart-methods

Warriorrrr reviewed (approved):
I approve of this

Doc94 committed:
Merge branch 'main' of github.com:PaperMC/Paper into feature/creaking-heart-methods

kennytv reviewed (approved):
It's so good that I need to approve again
